File: solutions/typescript/src/VendingMachine/VendingMachineState/ItemSelectedState.ts
import { Coin } from "../Coin";
import VendingMachine from "../VendingMachine";
import VendingMachineState from "./VendingMachineState";

export default class ItemSelectedState implements VendingMachineState {
    private machine: VendingMachine;

    constructor(machine: VendingMachine) {
        this.machine = machine;
    }

    insertCoin(coin: Coin) {
        const price = this.machine.getSelectedItem()?.getPrice();
        if (!price) {
            throw new Error("Item Not Selected Yet.");
        }

        this.machine.addBalance(coin);
        console.log("Coin Inserted: " + coin);

        if (this.machine.getBalance() >= price) {
            console.log("Sufficient money received.");
            this.machine.setState(this.machine.getHasMoneyState());
        }
    }

    selectItem() {
        console.log("Item already selected.");
    }

    dispense() {
        console.log("Please insert sufficient money.");
    }

    refund() {
        this.machine.resetToIdleState();
        this.machine.setState(this.machine.getIdleState());
    }
}

File: solutions/java/src/vendingmachine/state/ItemSelectedState.java
package vendingmachine.state;

import vendingmachine.enums.Coin;
import vendingmachine.VendingMachine;

public class ItemSelectedState extends VendingMachineState {
    public ItemSelectedState(VendingMachine machine) {
        super(machine);
    }

    @Override
    public void insertCoin(Coin coin) {
        machine.addBalance(coin.getValue());
        System.out.println("Coin Inserted: " + coin.getValue());
        int price = machine.getSelectedItem().getPrice();
        if (machine.getBalance() >= price) {
            System.out.println("Sufficient money received.");
            machine.setState(new HasMoneyState(machine));
        }
    }

    @Override
    public void selectItem(String code) {
        System.out.println("Item already selected.");
    }

    @Override
    public void dispense() {
        System.out.println("Please insert sufficient money.");
    }

    @Override
    public void refund() {
        machine.reset();
        machine.setState(new IdleState(machine));
    }
}
